packages/react/src/components/Toolbar/Toolbar.tsx

```diff
@@ -118,189 +97,348 @@ class Toolbar extends UIComponent<WithAsProp<ToolbarProps>, ToolbarState> {
-            overrideProps: itemOverridesFn,
-          })
-        case 'custom':
-          return ToolbarCustomItem.create(item, { overrideProps: itemOverridesFn })
-        default:
-          return ToolbarItem.create(item, { overrideProps: itemOverridesFn })
-      }
-    })
+    return _.map(
+      items,
+      (item: ShorthandValue<ToolbarItemProps & { kind?: ToolbarItemShorthandKinds }>) => {
+        const kind = _.get(item, 'kind', 'item')
+
+        switch (kind) {
+          case 'divider':
+            return ToolbarDivider.create(item, { overrideProps: itemOverridesFn })
+          case 'group':
+            return ToolbarRadioGroup.create(item, { overrideProps: itemOverridesFn })
+          case 'toggle':
+            return ToolbarItem.create(item, {
+              defaultProps: { accessibility: toggleButtonBehavior },
+              overrideProps: itemOverridesFn,
+            })
+          case 'custom':
+            return ToolbarCustomItem.create(item, { overrideProps: itemOverridesFn })
+          default:
+            return ToolbarItem.create(item, { overrideProps: itemOverridesFn })
+        }
+      },
+    )
   }
 
-  static getDerivedStateFromProps(props: ToolbarProps, state: ToolbarState) {
-    if (props.items === state.initialItems) {
-      return null
+  hide(el: HTMLElement) {
+    if (el.style.visibility === 'hidden') {
+      return
     }
 
-    return {
-      initialItems: props.items,
-      currentItems: props.items,
-      stableItems: state.stableItems,
-      stable: false,
+    el.style.visibility = 'hidden'
+    const wasFocusable = el.getAttribute('data-is-focusable')
+    if (wasFocusable) {
+      el.setAttribute('data-sd-was-focusable', wasFocusable)
     }
+    el.setAttribute('data-is-focusable', 'false')
   }
 
-  componentDidMount() {
-    this.afterComponentRendered()
+  show(el: HTMLElement) {
+    if (el.style.visibility !== 'hidden') {
+      return false
+    }
+
+    el.style.visibility = null
+    const wasFocusable = el.getAttribute('data-sd-was-focusable')
+    if (wasFocusable) {
+      el.setAttribute('data-is-focusable', wasFocusable)
+      el.removeAttribute('data-sd-was-focusable')
+    } else {
+      el.removeAttribute('data-is-focusable')
+    }
+
+    return true
   }
 
-  componentDidUpdate() {
-    this.afterComponentRendered()
+  /**
+   * Checks if `item` overflows a `container`.
+   * TODO: check and fix all margin combination
+   */
+  isItemOverflowing(itemBoundingRect: ClientRect, containerBoundingRect: ClientRect) {
+    return (
+      itemBoundingRect.right > containerBoundingRect.right ||
+      itemBoundingRect.left < containerBoundingRect.left
+    )
+  }
+
+  /**
+   * Checks if `item` would collide with eventual position of `overflowItem`.
+   */
+  wouldItemCollide(
+    $item: Element,
+    itemBoundingRect: ClientRect,
+    overflowItemBoundingRect: ClientRect,
+    containerBoundingRect: ClientRect,
+  ) {
+    let wouldCollide
+    if (this.rtl) {
```

Comment:
```suggestion
    if (this.props.rtl) {
```

Should it come from props?

Reply:
No, `rtl` is received as part of config in `renderComponent`, there I store it to `this`.

Reply:
Ouch, you're right 😭